### PR TITLE
[Merged by Bors] - prune data and compact state

### LINKED ISSUE #3049: prune proposals from database
## Description

proposals are ephemeral. once a unified content block is generated, they serve no practical purpose.

### LINKED ISSUE #3588: certificate: prune from network after hdist layer
## Description

depends on #2921

when a new node is syncing in the network, it still needs to sync hare certificate for every layer since genesis, because the last layer in tortoise is the last synced layer. 

we want to be able to prune certificate after for all layers before current - hdist when a node is syncing.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
Closes #3049 
Closes #3588 

## Changes
- prune data consistently with tortoise hdist distance. 
  - proposals
  - certificates
  - proposal id<->tid mapping
  
  certificate is needed for consensus within hdist.
  the same distance is used for all for simplicity.

- extract active set data from ballots and save them in activesets table.
- vacuum and checkpoint database after migration 4 is complete

this PR concludes the first update described https://github.com/spacemeshos/go-spacemesh/issues/4984#issuecomment-1720169998

### CODE REVIEW

File: sql/vacuum/vacuum.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	if _, err := db.Exec("vacuum", nil, nil); err != nil {
+		return fmt.Errorf("vacuum %w", err)
+	}
+	if _, err := db.Exec("pragma wal_checkpoint(TRUNCATE)", nil, nil); err != nil {
+		return fmt.Errorf("wal checkpoint %w", err)
+	}
+	return nil
+}
+
+func FreePct(db sql.Executor) (int, error) {
```
Comment (dshulyak) [inline sql/vacuum/vacuum.go]: could you please use full word? there is no reason to decipher this

File: mesh/janitor.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+				)
+			}
+			logger.Info("certificates pruned", zap.Stringer("lid", oldest))
+			if err := transactions.DeleteProposalTxs(db, oldest); err != nil {
+				logger.Error("failed to delete proposal tx mapping",
+					zap.Stringer("lid", oldest),
+					zap.Error(err),
+				)
+			}
+			logger.Info("proposal tx pruned", zap.Stringer("lid", oldest))
```
Comment (dshulyak) [inline mesh/janitor.go]: i think this is not useful as info

File: mesh/janitor.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		zap.Uint32("dist", safeDist),
+		zap.Duration("interval", interval),
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+			oldest := lc.CurrentLayer() - types.LayerID(safeDist)
+			if err := proposals.Delete(db, oldest); err != nil {
```
Comment (dshulyak) [inline mesh/janitor.go]: one thing to consider is that database is locked for the duration of the delete. so all other writes will be delayed. for example if delete takes 1s, all latency for handling ballot writes will be increased.

we should add histogram for total duration, and maybe split writes further, by deleting layer by layer

this is very important not to screw up ballots/proposals propagation

File: sql/vacuum/vacuum.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package vacuum
+
+import (
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+func VacuumDB(db sql.Executor) error {
```
Comment (dshulyak) [inline sql/vacuum/vacuum.go]: naming it Vacuum will be sufficient

File: cmd/root.go [hunk 5]
```diff
@@ -78,6 +78,10 @@ func AddCommands(cmd *cobra.Command) {
 		cfg.DatabaseConnections, "configure number of active connections to enable parallel read requests")
 	cmd.PersistentFlags().BoolVar(&cfg.DatabaseLatencyMetering, "db-latency-metering",
 		cfg.DatabaseLatencyMetering, "if enabled collect latency histogram for every database query")
+	cmd.PersistentFlags().IntVar(&cfg.DatabaseCompactStatePct, "db-compact-state-pct",
+		cfg.DatabaseCompactStatePct, "configure the threshold to trigger state compaction on startup")
+	cmd.PersistentFlags().DurationVar(&cfg.DatabasePruneInterval, "db-prune-interval",
+		cfg.DatabasePruneInterval, "configure interval for database pruning")
```
Comment (dshulyak) [inline cmd/root.go:82]: i am not sure if we need flags for everything. for example db-compact-state-pct is arcane parameter, that likely will be never changed

File: node/node.go [hunk 6]
```diff
@@ -169,6 +170,12 @@ func GetCommand() *cobra.Command {
 					return err
 				}
 
+				if app.Config.DatabaseCompactStatePct > 0 {
```
Comment (dshulyak) [inline node/node.go]: i realized that running vacuum will be too destructive if database size is even 20GB

we should either default to zero, or add an option that always runs compaction on startup and exit. so it is basically used as a tool

also i think there is no pressure to do it automatically, if code cleans up after itself continuously. what i think we should do is to execute migration that:
- drops proposals and txs
- nulls certificates
- update ballot to have nil active set
- run vacuum and checkpoint

once this is done, ephemeral state will not grow too much and will be reused for new data after old is cleaned up, so vacuum should be unnecessary

File: cmd/root.go (hunk 5 above)
Comment (countvonzero) [reply cmd/root.go:82]: removed

File: sql/migrations/0004_next.sql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+DELETE FROM proposals WHERE layer < 19000;
+DELETE FROM proposal_transactions WHERE layer < 19000;
+UPDATE certificates SET cert = NULL WHERE layer < 19000;
+UPDATE ballots SET ballot = prune_actives(ballot);
```
Comment (countvonzero) [inline sql/migrations/0004_next.sql]: considering #5002 and the fact that right now the ref ballot is still saved with activeset. 
`ActiveSetFromEpochFirstBlock` also still retrieve activeset directly from ref ballot.
is there a safe range i should apply to remove the activeset from ballots?

at layer 17800, the state size changed from 8.2GB to 470MB with this migration. the migration took 1.5 min.

File: sql/vacuum/vacuum.go (hunk 1 above)
Comment (countvonzero) [reply sql/vacuum/vacuum.go]: deleted.

File: node/node.go (hunk 6 above)
Comment (countvonzero) [reply node/node.go]: removed the automatic part.

Comment (countvonzero) [reply node/node.go]: please see https://github.com/spacemeshos/go-spacemesh/pull/4998#discussion_r1325486621

File: sql/database.go [hunk 8]
```diff
@@ -128,13 +139,11 @@ func Open(uri string, opts ...Opt) (*Database, error) {
 		if err != nil {
 			return nil, err
 		}
-	}
-	for i := 0; i < config.connections; i++ {
-		conn := pool.Get(context.Background())
-		if err := registerFunctions(conn); err != nil {
-			return nil, err
+		if before == 3 {
+			if err := Vacuum(db); err != nil {
```
Comment (countvonzero) [inline sql/database.go:153]: vaccum needs to happen outside of the db transaction, so the vaccum/checkpoint cannot be part of the migration.

File: sql/migrations/0004_next.sql (hunk 7 above)
Comment (dshulyak) [reply sql/migrations/0004_next.sql]: my proposal was to save active sets into activeset table while pruning. and then we either have an option to sql-join them on activesethash field (requires extending ballots table so maybe not optimal) or do another query to get activeset hash when we need it.

File: sql/migrations/0004_next.sql [hunk 9]
```diff
@@ -0,0 +1,5 @@
+DELETE FROM proposals WHERE layer < 19000;
+DELETE FROM proposal_transactions WHERE layer < 19000;
```
Comment (dshulyak) [inline sql/migrations/0004_next.sql]: i think this we can delete regardless of layer

File: sql/migrations/0004_next.sql (hunk 7 above)
Comment (dshulyak) [reply sql/migrations/0004_next.sql]: > https://github.com/spacemeshos/go-spacemesh/issues/4984#issuecomment-1717409692

we should also stop writing active sets encoded together with ballot, after this migration is done. not necessarily within same change, but would be better to release them together

Comment (countvonzero) [reply sql/migrations/0004_next.sql]: changed to extract active set from ballots and save the prune blob.

please consider https://github.com/spacemeshos/go-spacemesh/issues/4984#issuecomment-1720169998
i'll continue with the other code change accordingly.

File: mesh/janitor.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/spacemeshos/go-spacemesh/mesh/metrics"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/activesets"
+	"github.com/spacemeshos/go-spacemesh/sql/ballots"
+	"github.com/spacemeshos/go-spacemesh/sql/certificates"
+	"github.com/spacemeshos/go-spacemesh/sql/proposals"
+	"github.com/spacemeshos/go-spacemesh/sql/transactions"
+)
+
+func Prune(
```
Comment (dshulyak) [inline mesh/janitor.go]: why is it in mesh module? lets please create another one, which is called pruner (or janitor, though pruner is more specific). maybe it makes sense to be in sql/pruner

File: mesh/janitor.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+					zap.Stringer("lid", oldest),
+					zap.Error(err),
+				)
+			}
+			metrics.PrunePropTxLatency.Observe(time.Since(t2).Seconds())
+		}
+	}
+}
+
+func ExtractActiveSet(db sql.Executor) error {
```
Comment (dshulyak) [inline mesh/janitor.go]: this is only used for migrations? can we put it somewhere that is not in the consensus code, maybe ballots/util/extract.go

**Review by dshulyak — APPROVED** [review]
please see notes about pruner/extract code

File: mesh/janitor.go (hunk 10 above)
Comment (countvonzero) [reply mesh/janitor.go]: moved to pruner module.

File: mesh/janitor.go (hunk 11 above)
Comment (countvonzero) [reply mesh/janitor.go]: put it in sql/ballots/util/extract.go.

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**